roachprod-microbench: list benchmarks from repo

herkolategan opened the pull request:
Currently, the microbenchmark job uses the `go -test.list` functionality on each test binary to get a full list of the available microbenchmarks to run. This has worked well, but we want to add some new functionality that allows microbenchmakrs to determine a run configuration from the `doc` or `comment` on the microbenchmark function.

This change is the first step to have a list functionality to list microbenchmarks via the repo rather than the binaries and to also parse the doc for an additional run configuration.

Epic: None
Release Note: None

herkolategan commented:
Example output:
```
# ./bin/roachprod-microbench list
BenchmarkSpansForAllTableIndexes [disaster-recovery]
BenchmarkCoverageChecks [disaster-recovery]
BenchmarkRestoreEntryCover [disaster-recovery]
BenchmarkIteratorMemory [disaster-recovery]
BenchmarkPrefixRewriter [disaster-recovery]
BenchmarkSelect1 [sql-queries]
BenchmarkSelect2 [sql-queries]
BenchmarkSelect3 [sql-queries]
...
```

This is the first PR for allowing specific microbenchmark configurations. A follow-up PR will export this output to a file that will be consumed during the microbenchmarks TC job to determine which microbenchmarks to run, including a configuration if specified by the benchmark.

herkolategan committed:
roachprod-microbench: list benchmarks from repo

Currently, the microbenchmark job uses the `go -test.list` functionality on each
test binary to get a full list of the available microbenchmarks to run. This has
worked well, but we want to add some new functionality that allows
microbenchmakrs to determine a run configuration from the `doc` or `comment` on
the microbenchmark function.

This change is the first step to have a list functionality to list
microbenchmarks via the repo rather than the binaries and to also parse the doc
for an additional run configuration.

Epic: None
Release Note: None

herkolategan committed:
roachprod-microbench: list command

Add a CLI command to output the benchmark list.

Epic: None
Release Note: None

srosenberg commented on pkg/cmd/roachprod-microbench/list.go:
```diff
@@ -0,0 +1,182 @@
+const docMarker = "benchmark-ci:"
+
+func listPackages(pkgDir string, patterns ...string) ([]*packages.Package, error) {
+	pkgs, err := packages.Load(
```
This could take a while... wonder if it'd be faster just to `WalkDir` over all `_test.go` and `parseFile`, since you don't really need types or imports.

herkolategan commented on pkg/cmd/roachprod-microbench/list.go:
```diff
@@ -0,0 +1,177 @@
+		if !ok || fd.Recv != nil || fd.Name == nil || !strings.HasPrefix(fd.Name.Name, "Benchmark") {
+			return true
+		}
+		if len(fd.Type.Params.List) == 0 {
```
True, tightened up the requirements.

herkolategan committed:
roachprod-microbench: concurrently analyze benchmarks

Performance improvement to analyze files for benchmarks in parallel.

Epic: None
Release Note: None

srosenberg commented on pkg/cmd/roachprod-microbench/list.go:
```diff
@@ -0,0 +1,206 @@
+
+	var benchmarkInfoList []BenchmarkInfo
+	g := errgroup.Group{}
+	resultsCh := make(chan []BenchmarkInfo)
```
Could make this buffered with a reasonably conservative capacity, e.g., `runtime.NumCPU()`.

herkolategan commented:
bors r+
